File: Completed-Algorithms/Leetcode_739_Daily_Temperatures/monostack.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    const stack = []
    const answers = new Array(temperatures.length).fill(0)

    for(let i = 0; i < temperatures.length; i++) {
        while(stack.length > 0 && temperatures[i] > stack[stack.length - 1].temp) {
            const {prevIndex} = stack[stack.length - 1]
            answers[prevIndex] = i - prevIndex
            stack.pop()
        }
        stack.push({prevIndex: i, temp: temperatures[i]})
    }
    return answers
};

File: Completed-Algorithms/Leetcode_739_Daily_Temperatures/monostack.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        answer = [0] * len(temperatures)
        stack = []

        for index, t in enumerate(temperatures):
            while stack and t > temperatures[stack[-1]]:
                old_index = stack.pop()
                answer[old_index] = index - old_index
            stack.append(index)
        return answer
